pkg/model/utxo/dust_db.go

```diff
@@ -0,0 +1,223 @@
+	if err != nil {
+		return err
+	}
+
+	newDustAllowanceBalance := int64(dustAllowanceBalance) + dustAllowanceBalanceDiff
+	newDustOutputCount := dustOutputCount + dustOutputCountDiff
+
+	if newDustOutputCount < 0 || newDustAllowanceBalance < 0 {
+		// Count or balance cannot be negative
+		return fmt.Errorf("%w: %s dustAllowanceBalance %d, dustOutputCount %d", ErrInvalidDustForAddress, hex.EncodeToString(addressBytes), dustAllowanceBalance, dustOutputCount)
+	}
+
+	return u.storeDustForAddress(addressBytes, uint64(newDustAllowanceBalance), newDustOutputCount, mutations)
+}
+
+func (u *Manager) applyNewDustWithoutLocking(newOutputs Outputs, newSpents Spents) error {
+	dustDiff := make(map[string]*DustDiff)
+
+	for _, output := range newOutputs {
+		// Add new dust
+		switch output.outputType {
+		case iotago.OutputSigLockedDustAllowanceOutput:
+			address := output.Address().String()
+			diff, found := dustDiff[address]
+			if found {
+				diff.DustAllowanceBalanceDiff += int64(output.Amount())
+			} else {
+				dustDiff[address] = NewDustDiff(int64(output.Amount()), 0)
+			}
+		case iotago.OutputSigLockedSingleOutput:
+			if output.Amount() < iotago.OutputSigLockedDustAllowanceOutputMinDeposit {
+				address := output.Address().String()
+				diff, found := dustDiff[address]
+				if found {
+					diff.DustOutputCount += 1
+				} else {
+					dustDiff[address] = NewDustDiff(0, 1)
+				}
+			}
+		}
+	}
+
+	for _, spent := range newSpents {
+
+		// Remove spent dust
+		output := spent.Output()
+		switch output.outputType {
+		case iotago.OutputSigLockedDustAllowanceOutput:
+			address := output.Address().String()
+			diff, found := dustDiff[address]
+			if found {
+				diff.DustAllowanceBalanceDiff -= int64(output.Amount())
+			} else {
+				dustDiff[address] = NewDustDiff(-int64(output.Amount()), 0)
+			}
+		case iotago.OutputSigLockedSingleOutput:
+			if output.Amount() < iotago.OutputSigLockedDustAllowanceOutputMinDeposit {
+				address := output.Address().String()
+				diff, found := dustDiff[address]
+				if found {
+					diff.DustOutputCount -= 1
+				} else {
+					dustDiff[address] = NewDustDiff(0, -1)
+				}
+			}
+		}
+	}
+
+	return u.applyDustDiff(dustDiff)
+}
+
+func (u *Manager) rollbackDustWithoutLocking(newOutputs Outputs, newSpents Spents) error {
+
+	dustDiff := make(map[string]*DustDiff)
+
+	// we have to delete the newOutputs of this milestone
+	for _, output := range newOutputs {
+
+		// Remove unspent dust
+		switch output.outputType {
+		case iotago.OutputSigLockedDustAllowanceOutput:
+			address := output.Address().String()
+			diff, found := dustDiff[address]
+			if found {
+				diff.DustAllowanceBalanceDiff -= int64(output.Amount())
+			} else {
+				dustDiff[address] = NewDustDiff(-int64(output.Amount()), 0)
+			}
+		case iotago.OutputSigLockedSingleOutput:
+			if output.Amount() < iotago.OutputSigLockedDustAllowanceOutputMinDeposit {
+				address := output.Address().String()
```

Comment:
you should convert the address bytes to string, but the String function uses hex encoding